### PR TITLE
Add support for gc2gde and gd2gce erfa functions in units.

### PR DESCRIPTION
**Pull request by cmarmo**
### Description
This pull request adds support for `gc2gde` and `gd2gce` erfa functions in `units`.
Tests are added.

Follows the suggestion in https://github.com/astropy/astropy/pull/14714#pullrequestreview-1407107055.

Towards https://github.com/astropy/astropy/issues/11170.

Thanks for considering it. 

This work is funded by the Europlanet 2024 Research Infrastructure (RI) Grant.

### CODE REVIEW

**Comment by cmarmo** [comment]
Note that I can't reproduce the coverage failure locally ... any suggestion is welcome! Thanks!

**Comment by pllim** [comment]
The coverage failure was because coverage report upload fails. Happens once in a while.

Still needs a change log. Thanks!

File: astropy/units/quantity_helper/erfa.py [hunk 1]
```diff
@@ -98,6 +99,22 @@ def helper_gc2gd(f, nounit, unit1):
         )
 
 
+def helper_gc2gde(f, unit1, dimensionless_unscaled, unit2):
```
Comment (mhvk) [inline astropy/units/quantity_helper/erfa.py:102]: Like below, enumerate all units, and have `get_converter(unit2, dimensionless_unscaled)` below.

By the way, feel free to give the units more meaningful names - I know I didn't do it for many functions but I changed later (see, e.g., `helper_s2pv`).

**Review by mhvk — COMMENTED** [review]
Thanks! A few comments on the implementation...

File: astropy/units/quantity_helper/erfa.py [hunk 2]
```diff
@@ -98,6 +99,24 @@ def helper_gc2gd(f, nounit, unit1):
         )
 
 
+def helper_gc2gde(f, unit_r, unit_flat, unit_xyz):
+    from astropy.units.si import m, radian
+
+    if unit_r is None:
```
Comment (mhvk) [inline astropy/units/quantity_helper/erfa.py:105]: I don't think this is necessary - just let `get_converter` handle this!

I realize the error messages may not be ideal, but I think it may be better in that case to try to improve `get_converter` -- but in a different PR!

File: astropy/units/quantity_helper/erfa.py [hunk 3]
```diff
@@ -117,6 +136,21 @@ def helper_gd2gc(f, nounit, unit1, unit2, unit3):
         )
 
 
+def helper_gd2gce(f, unit_r, unit_flat, unit_long, unit_lat, unit_h):
+    from astropy.units.si import m, radian
+
+    if unit_r is None:
```
Comment (mhvk) [inline astropy/units/quantity_helper/erfa.py:142]: Same, just keep it simple!

File: astropy/units/quantity_helper/erfa.py [hunk 4]
```diff
@@ -98,6 +99,24 @@ def helper_gc2gd(f, nounit, unit1):
         )
 
 
+def helper_gc2gde(f, unit_r, unit_flat, unit_xyz):
+    from astropy.units.si import m, radian
+
+    if unit_r is None:
+        raise UnitTypeError("Equatorial radius should have length units.")
+    unit_flat = unit_r / unit_r
```
Comment (mhvk) [inline astropy/units/quantity_helper/erfa.py:107]: Why is this necessary? You are overriding the input unit (which might be, say, `u.percent`, which could be converted to `dimensionless_unscaled`)

File: astropy/units/tests/test_quantity_erfa_ufuncs.py [hunk 5]
```diff
@@ -504,3 +510,93 @@ def test_apio(self):
...
+            )
+        with pytest.raises(UnitTypeError, match=msg):
+            xyz, status = erfa_ufunc.gd2gce(
+                self.equatorial_radius_value,
+                self.flattening,
+                self.lon,
+                self.lat,
+                self.height,
+            )
+        msg = "'NoneType' object has no attribute '_get_converter'"
```
Comment (mhvk) [inline astropy/units/tests/test_quantity_erfa_ufuncs.py:554]: That's actually a pretty terrible error message! I think I definitely see a point now of trying to do better - but let's do so in a different PR!

**Review by mhvk — COMMENTED** [review]
Sorry, a few more comments. Some noting that really the existing code can use a bit of improvement...  Sigh...

File: astropy/units/quantity_helper/erfa.py (hunk 4 above)
Comment (cmarmo) [reply astropy/units/quantity_helper/erfa.py:107]: ok, understood: this was the reason of the error on `unit_r` being managed separately.
I'm going to remove it then the check on `unit_r` will go too.

**Comment by cmarmo** [comment]
@mhvk I just realized that the `erfa.eform` function (which I need to retrieve the ellipsoid parameters from the standard string, when addressing the [item .2.](https://github.com/astropy/astropy/pull/14714#pullrequestreview-1407107055)) does not have its own helper.
I'm going to add it here too if this is ok for you.

File: astropy/units/quantity_helper/erfa.py [hunk 6]
```diff
@@ -85,6 +86,12 @@ def helper_p2s(f, unit1):
     return [None], (radian, radian, unit1)
 
 
+def helper_eform(f, nounit):
+    if nounit is not None:
+        raise UnitTypeError("Ellipsoid cannot be a quantity.")
+    return [None], (None, None)
```
Comment (mhvk) [inline astropy/units/quantity_helper/erfa.py:92]: Good idea to add coverage for `eform` too! But this doesn't seem right: the equatorial radius and flattening do have units: `(m, dimensionless_unscaled)` (and `from astropy.units.si import m` above)

Comment (cmarmo) [reply astropy/units/quantity_helper/erfa.py]: That was my first try... I couldn't make it work.
No matter how I changed the output dimensions, `eform` always gives me a-dimensional outputs... 
I have pushed the "dimensional" version for you to check... I'm probably missing something.

Comment (mhvk) [reply astropy/units/quantity_helper/erfa.py]: Ah, it actually makes sense - I should have thought of this! The overrides only work if one puts in a `Quantity`as input (or output). Without that, the ufunc has no way to know a `Quantity` is expected and therefore just acts as if it regular `ndarray` input. We can actually not change this in astropy in any simple way.

In principle, the helper function will only be called if you put in a `Quantity` to hold the *output*. This is not in itself bad, but also not super useful. Perhaps the simplest right now is to just remove this again from the PR, and then in the next PR just work with the raw version, i.e., call it and then turn the output into `Quantity` there.

Comment (cmarmo) [reply astropy/units/quantity_helper/erfa.py]: > Perhaps the simplest right now is to just remove this again from the PR, and then in the next PR just work with the raw version, i.e., call it and then turn the output into Quantity there.

Thanks for your explanation! Makes sense... :)
Just to be sure: you mean I should go back to https://github.com/astropy/astropy/pull/14729/commits/34b642fa9ae3669adf411c8e9a6551775a91a1f9 or completely remove `eform` from the helpers? Thanks!

**Comment by cmarmo** [comment]
I have removed `eform` from this pull request, synchronized with main and squashed the commits.
Thanks for your patience.

File: astropy/units/tests/test_quantity_erfa_ufuncs.py [hunk 7]
```diff
@@ -504,3 +509,93 @@ def test_apio(self):
...
+        self.y = self.y_value << self.length_unit
+        self.z_value = 5.244e6
+        self.z = self.z_value << self.length_unit
+
+    def test_unit_errors(self):
+        """Test unit errors when dimensionless parameters are used"""
+
+        msg = "'NoneType' object has no attribute '_get_converter'"
+        xyz = np.array([self.x_value, self.y_value, self.z_value]) << self.length_unit
+        status = 0
```
Comment (mhvk) [inline astropy/units/tests/test_quantity_erfa_ufuncs.py:541]: No need to set `status = 0` beforehand, since you're not passing it in. Indeed, since we're just looking for the error message, you can just call the `ufunc` without assigning to any outputs.

File: astropy/units/tests/test_quantity_erfa_ufuncs.py [hunk 8]
```diff
@@ -504,3 +509,93 @@ def test_apio(self):
...
+        self.y_value = 3e6
+        self.y = self.y_value << self.length_unit
+        self.z_value = 5.244e6
+        self.z = self.z_value << self.length_unit
+
+    def test_unit_errors(self):
+        """Test unit errors when dimensionless parameters are used"""
+
+        msg = "'NoneType' object has no attribute '_get_converter'"
+        xyz = np.array([self.x_value, self.y_value, self.z_value]) << self.length_unit
```
Comment (mhvk) [inline astropy/units/tests/test_quantity_erfa_ufuncs.py:540]: Real nitpick, but might as well do `xyz = np.stack([self.x, self.y, self.z])`.
Indeed, since you use it multiple times, might as well define it in `setup_class`

File: astropy/units/tests/test_quantity_erfa_ufuncs.py [hunk 9]
```diff
@@ -504,3 +509,93 @@ def test_apio(self):
...
+            )
+        with pytest.raises(AttributeError, match=msg):
+            xyz, status = erfa_ufunc.gd2gce(
+                self.equatorial_radius_value,
+                self.flattening,
+                self.lon,
+                self.lat,
+                self.height,
+            )
+        xyz = np.array([self.x_value, self.y_value, self.z_value])
```
Comment (mhvk) [inline astropy/units/tests/test_quantity_erfa_ufuncs.py:554]: Continuing with nitpicks, how about removing this and instead do `xyz.value` in the call below?

**Review by mhvk — APPROVED** [review]
No, really, thanks for *your* patience!  I have some final nitpicks, but am approving already, since things look good!

**Comment by cmarmo** [comment]
nitpicks addressed... :)

**Review by mhvk — APPROVED** [review]
OK, all good now - let get it in! Thanks!!

**Comment by cmarmo** [comment]
Wonderful! Thanks! :hugs: :partying_face:

**Comment by pllim** [comment]
From the sidelines, I really appreciate all these non-trivial fixes both in FITS and coordinates, @cmarmo ! Looking forward to more. 👏

**Comment by pllim** [comment]
> Europlanet 2024 Research Infrastructure (RI) Grant

This is amazing! I am so glad to see more grants agreeing to invest dev time into the Project. cc @astropy/coordinators

**Comment by cmarmo** [comment]
> This is amazing! I am so glad to see more grants agreeing to invest dev time into the Project

This was possible thanks to @Erard and the involvement of the _Observatoire de Paris_ in the Europlanet project. :pray:

**Comment by hamogu** [comment]
> This work is funded by the Europlanet 2024 Research Infrastructure (RI) Grant.

It's great that funded projects contribute to astropy. That's exactly how Open source is supposed to work and we welcome your contributions. I'm just commenting here to let you know that you of course you are also welcome to connect on [slack](https://astropy.slack.com/), on the [astropy-dev mailing list](http://groups.google.com/group/astropy-dev) or by email to the [Coordination Committee](mailto:coordinators@astropy.org) if you ever need something "official looking" for e.g. a grant proposal.

**Comment by cmarmo** [comment]
Thanks @hamogu!
Apparently I cannot create an account on slack as I don´t have the right domain... :)
I may need an invitation? Thanks!

**Comment by hamogu** [comment]
Maybe the slack link expired. I'll check with the slack admins and get back to you.

**Comment by hamogu** [comment]
Can you try: http://joinslack.astropy.org/ ?

**Comment by cmarmo** [comment]
That worked! Thanks a lot!

> Maybe the slack link expired. 

...my bad for being off-line for some time...

**Comment by hamogu** [comment]
I think I just linked the wrong link in https://github.com/astropy/astropy/pull/14729#issuecomment-1587669497
Since I'm signed up already, I did not notice the difference between sign-in and sign-up. Sorry ;-)
But I'm glad it worked!
